// es_client/core_plugins/base_plugin/base_plugin.js
if (typeof define !== 'function') { var define = require('amdefine')(module) }
define( function(require,exports,module) {
    const PluginsManager = require('lib/plugins_manager').PluginsManager;

    class BasePlugin {
        constructor(config) {
            this.config          = config;
            this.name            = this.config.NAME;
            this.plugins_manager = new PluginsManager();
            this.view            = null;
        }

        subscribeEvents(){
            this.plugins_manager.subscribeCreateViewEvent(this.createView.bind(this));
        }

        createView(params, callback)
        {
            require([this.plugins_manager.getPluginPath(this.config.NAME) + '/views/' + this.config.VIEW],
                function(View)
                {
                    params.config = this.config;
                    this.view = new View.View(params);
                    this.view.preRendering(callback);
                }.bind(this)
            );
        }
    };

    module.exports = new BasePlugin(require('./config'));
    module.exports.BasePlugin = BasePlugin;
});
