// es_client/core_plugins/function_menu/views/function_menu_view.js
if (typeof define !== 'function') { var define = require('amdefine')(module) }
define( function(require,exports,module)
{
    const ExpressionHelpers = require('../../../lib/expression_helpers');

    const BaseView = require('core_plugins/base_plugin/views/base_view').View;

    class FunctionMenuView extends BaseView {
        constructor(params) {
            super(params);
            this.config = JSON.parse(JSON.stringify(this.config));
            this.config.PLACEHOLDER = "es_panel_0"
            this.config.ELEMENT     = "#es-function-menu-container";
            this.template_view      = "function_menu_view"
            this.events = {
                'click .es-menu-button': 'onButtonClick'
            };
            this.template_data = {eh : ExpressionHelpers(this.es_controller.data)};
        }

        preRendering()
        {
            this.render();
        }

        getCurrentCell(){
            let cells  = this.models.get('selection').getCells();//this.getSelection().getCells();
            if(!cells) return;
            return cells[0];
        }

        onButtonClick(e){
            let action = "=" + $(e.currentTarget).data('action');
            let cell   = this.getCurrentCell();
            let sheet  = this.getSheet();
            sheet.updateCell(cell.row_id, cell.col_id, action);
            var input_selector = "#" + cell.row_id + '-' + cell.col_id + '-input';
            $(input_selector).focus();
        }

        onButtonClick(e){
            let action = "=" + $(e.currentTarget).data('action');
            let cell   = this.getCurrentCell();
            let sheet  = this.getSheet();
            sheet.updateCell(cell.row_id, cell.col_id, action);
            var input_selector = "#" + cell.row_id + '-' + cell.col_id + '-input';
            $(input_selector).focus();
        }
    }

    module.exports.View = FunctionMenuView;
});
