// es_client/plugins/csv_manager/views/csv_manager_view.js
if (typeof define !== 'function') {
    var define = require('amdefine')(module);
}
define( function(require,exports,module) {
    const BaseView         = require('core_plugins/base_plugin/views/base_view').View;
    const TableViewEvents  = require('core_plugins/table/views/table_view').Events;

    const ImportDialogView = require('./csv_manager_import_dialog_view').View;

    class CsvManagerView extends BaseView {
        constructor(params) {
            super(params);
            this.config = JSON.parse(JSON.stringify(this.config));
            this.config.PLACEHOLDER = "es_panel_0"
            this.config.ELEMENT     = "#es-csv-manager-menu-container";
            this.template_view      = "csv_manager_view"
            this.events = {
                'click #es-import-csv' : 'onClickImportCsv'
            };
            this.initializeElements();

            this.importDialog    = new ImportDialogView(params);
            this.importDialog.el = $('#es-modal-box');
        }

        preRendering() {
            this.template_data = {current_sheet_id : this.es_controller.table.template_data.id};
            super.render();
        }

        onClickImportCsv(e)
        {
            this.importDialog.preRendering();
        }
    };

    module.exports.View = CsvManagerView;

});
