// es_client/plugins/maps/views/map_menu.js
if (typeof define !== 'function') { var define = require('amdefine')(module) }
define( function(require,exports,module)
{
    const BaseView        = require('core_plugins/base_plugin/views/base_view').View;
    const CellMenuView    = require('core_plugins/table/views/cell_menu_view').View;
    const MenuViewEvents  = require('core_plugins/table/views/menu_view').Events;
    const MapDialog       = require('./map_dialog').View;
    const MapPreviewView  = require('./map_preview_view').View;

    class MapMenu extends BaseView {
        constructor(params) {
            super(params);
            this.plugin_manager.subscriveEvent(MenuViewEvents.MenuViewEventsMenuRendering, this.addActionToCellMenu.bind(this));
            params.el         = null;
            params.cell       = null;

            this.mapDialog    = new MapDialog(params);
            this.mapDialog.el = $('#es-modal-box');
            this.mapDialog.loadStyle();

            this.mapPreview = new MapPreviewView(params);
        }

        addActionToCellMenu(e)
        {
            if(e.menu instanceof CellMenuView) {
                let menu    = e.menu;
                let actions = [];
                actions.push(
                    {
                        action: "add_geo_point",
                        callback: function () {
                            top.postMessage("open-select-merker-map_event", 'http://' + window.location.hostname);
                            this.mapDialog.cell = e.menu.current_element;
                            this.mapDialog.loadView();
                        }.bind(this)
                    }
                );
                this.plugin_manager.emitEvent(MenuViewEvents.MenuViewEventsAddActions, {actions : actions, menu_instance : e.menu});
            }
        }
    }

    module.exports.View = MapMenu;
});
