// es_client/plugins/upload_images/views/upload_image_dialog_view.js
if (typeof define !== 'function') {
    let define = require('amdefine')(module);
}
define( function(require,exports,module) {
    const $        = require('jquery');
    const BaseView = require('core_plugins/base_plugin/views/base_view').View;

    class UploadImageDialogView extends BaseView {
        constructor(params) {
            super(params);
            this.config = JSON.parse(JSON.stringify(this.config));
            this.events = {
                'click #upload_image_button': 'uploadImage',
                'dragover #dragzone'        : 'onFileDrag',
                'dragleave #dragzone'       : 'onFileDragLeave',
                'drop #dragzone'            : 'onFileDrop',
                'change #image_file'        : 'fileSelected',
            };
            this.initializeElements();
            this.template_view          = "upload_image_dialog_view";
            this.config.ELEMENT_TO_FILL = "#es-modal-box";
            this.config.ELEMENT         = "#es-modal-box";
        }

        initializeElements()
        {
            super.initializeElements();
            this.modal_overlay     = $('#es-modal-overlay');
            this.overlay           = $(".es-overlay");
            this.upload_image_form = $("#upload_image_form");
            this.dragzone          = $("#dragzone");
            this.dragzone_caption  = $("#dragzone-caption");
            this.image_file        = $("#image_file")[0];
        }

        preRendering() {
            this.template_data = {current_sheet_id : this.es_controller.table.template_data.id};
            this.modal_overlay.show();
            super.render();
        }

        uploadImage(){
            let self = this;
            this.upload_image_form
                .submit( function( e ) {
                    $.ajax( {
                        url: '/ethersheet/upload/image',
                        type: 'POST',
                        data: new FormData( this ),
                        processData: false,
                        contentType: false,
                        success : function(data) {
                            //udate cell value
                            self.cell.val(e.data);
                            self.es_controller.table.getSheet().updateCell(self.cell.data('row_id'), self.cell.data('col_id'), data.image_url);
                            //dismiss dialog and menu
                            self.cell.on('mouseover', function(e){self.es_controller.table.showCellPreview(e)}.bind(self));

                            self.modal_overlay.hide();
                            self.overlay.remove();
                        },
                        error : function(data){
                            alert(data.statusText + ": the file size must be less than 3Mb");
                            self.modal_overlay.hide();
                            self.overlay.remove();
                        }
                    } );
                    e.preventDefault();
                } );
        }

        onFileDrag(e){
            e.stopPropagation();
            e.preventDefault();
            this.dragzone.addClass("ondrag");
        }

        onFileDragLeave(e){
            this.dragzone.removeClass("ondrag");
        }

        onFileDrop(e){
            e.preventDefault();
            this.dragzone.removeClass("ondrag");

            let files;
            if(_.isUndefined(e.dataTransfer))
                files = e.originalEvent.dataTransfer.files;
            else
                files = e.dataTransfer.files;
            this.image_file.files = files;
            this.showImagePreview(files[0]);

        }

        fileSelected(){
            this.showImagePreview(this.image_file.files[0]);
        }

        showImagePreview(img){
            this.dragzone_caption.html("");
            if(!_.isUndefined(typeof FileReader)){
                let fs = new FileReader();
                fs.onload = function(e){
                    this.dragzone.html('<img src="' + e.target.result  + '" />');
                }.bind(this);
                fs.readAsDataURL(img);
            }else{
                this.dragzone_caption.html(files[0].name);
            }
        }
    };

    module.exports.View = UploadImageDialogView;

});
