// es_client/plugins/upload_images/views/upload_images_menu.js
if (typeof define !== 'function') { let define = require('amdefine')(module) }
define( function(require,exports,module)
{
    const BaseView         = require('core_plugins/base_plugin/views/base_view').View;
    const CellMenuView      = require('core_plugins/table/views/cell_menu_view').View;
    const MenuViewEvents    = require('core_plugins/table/views/menu_view').Events;
    const UploadImageDialog = require('./upload_image_dialog_view').View;
    const ImagePreviewView  = require('./image_preview_view').View;

    class UploadImageMenu extends BaseView {
        constructor(params) {
            super(params);
            this.plugin_manager.subscriveEvent(MenuViewEvents.MenuViewEventsMenuRendering, this.addActionToCellMenu.bind(this));
            params.el         = null;
            params.cell       = null;

            this.uploadImageDialog    = new UploadImageDialog(params);
            this.uploadImageDialog.el = $('#es-modal-box');

            this.imagePreview = new ImagePreviewView(params);
        }

        addActionToCellMenu(e)
        {
            if(e.menu instanceof CellMenuView) {
                let menu    = e.menu;
                let actions = [];
                actions.push(
                    {
                        action: "add_image",
                        callback: function () {
                            this.uploadImageDialog.cell = e.menu.current_element;
                            this.uploadImageDialog.preRendering();
                        }.bind(this)
                    }
                );
                this.plugin_manager.emitEvent(MenuViewEvents.MenuViewEventsAddActions, {actions : actions, menu_instance : e.menu});
            }
        }
    }

    module.exports.View = UploadImageMenu;
});
