// es_client/core_plugins/main_container/views/main_container_view.js
if (typeof define !== 'function') { var define = require('amdefine')(module) }
define( function(require,exports,module)
{
    const BaseView = require('core_plugins/base_plugin/views/base_view').View;

    class MainContainerView extends BaseView {
        constructor(params) {
            super(params);
            this.events = {
                'click #es-modal-close': 'closeModal',
                'click .es-sidebar-toggle': 'toggleSidebar',
            };
            this.$expression = null;
            this.$table = null;
            this.$grid = null;
            this.is_rendered = false;

            this.template_data = {title : ""};
        }

        initializeElements()
        {
            super.initializeElements();
            this.$expression        = $("#es-expression-editor-container");
            this.$table             = $("#es-table-container");
            this.$menu              = $("#es-menu-container");
            this.$panel_0           = $("#es-panel-0");
            this.$panel_1           = $("#es-panel-1");
            this.$modal_overlay     = $("#es-modal-overlay");

            this.is_rendered = true;
            this.resize();
        }

        preRendering(callback){
            super.render(callback);
        }

        resize(){
            if(!this.is_rendered) return;
        }

        closeModal(){
           this.$modal_overlay.hide()
        }

        toggleSidebar(e) {
            let speed = 400;
            let $el = $(e.currentTarget);

            this.$es_sidebar_toggle = $('.es-sidebar-toggle');
            this.$menu_container    = $(".menu-container");

            if ($el.hasClass('active')) {
                //close the panel
                $el.removeClass('active');
                this.$menu.animate({'right': '258px'}, speed);
                this.$panel_0.animate({'width': '0'}, speed, 'swing', function () {
                    this.$panel_0.hide();
                    this.$menu_container.hide();
                }.bind(this));
                this.$panel_1.animate({'margin-left': '10px'}, speed);
            } else {
                this.$es_sidebar_toggle.removeClass('active');
                $el.addClass('active');
                this.$panel_0.show();
                this.$menu_container.hide();
                var container = $el.attr('id').replace('icon', 'menu-container');
                $("#" + container).show();
                if (this.$panel_0.width() <= 0) {
                    this.$menu.animate({'right': '0px'}, speed);
                    this.$panel_0.animate({'width': '258px'}, speed);
                    this.$panel_1.animate({'margin-left': '255px'}, speed);
                }
            }
        }
    }

    module.exports.View = MainContainerView;
});
