// es_client/plugins/shape_file_converter/views/shape_file_converter_view.js
if (typeof define !== 'function') {
    var define = require('amdefine')(module);
}
define( function(require,exports,module) {
    const BaseView         = require('core_plugins/base_plugin/views/base_view').View;
    const TableViewEvents  = require('core_plugins/table/views/table_view').Events;

    const ImportDialogView = require('./shape_file_converter_import_dialog_view').View;

    class ShapeFileConverterView extends BaseView {
        constructor(params) {
            super(params);
            this.config = JSON.parse(JSON.stringify(this.config));
            this.config.PLACEHOLDER = "es_panel_0"
            this.config.ELEMENT     = "#es-shape-file-converter-menu-container";
            this.template_view      = "shape_file_converter_view"
            this.events = {
                'click #es-import-shape-file' : 'onClickImportShapeFile'
            };
            this.initializeElements();

            this.importDialog    = new ImportDialogView(params);
            this.importDialog.el = $('#es-modal-box');
        }

        preRendering() {
            this.template_data = {current_sheet_id : this.es_controller.table.template_data.id};
            super.render();
        }

        onClickImportShapeFile(e)
        {
            this.importDialog.preRendering();
        }
    };

    module.exports.View = ShapeFileConverterView;

});
