// es_client/plugins/shape_file_converter/views/shape_file_converter_import_dialog_view.js
if (typeof define !== 'function') {
    let define = require('amdefine')(module);
}
define( function(require,exports,module) {
    const $        = require('jquery');
    const BaseView = require('core_plugins/base_plugin/views/base_view').View;

    class ImportDialogView extends BaseView {
        constructor(params) {
            super(params);
            this.config = JSON.parse(JSON.stringify(this.config));
            this.events = {
               
            };
            this.initializeElements();
            this.template_view          = "shape_file_converter_import_dialog_view";
            this.config.ELEMENT_TO_FILL =  "#es-modal-box"
            this.config.ELEMENT         =  "#es-modal-box"
        }

        initializeElements()
        {
            super.initializeElements();
            this.modal_overlay = $('#es-modal-overlay');
            this.overlay       = $(".es-overlay");
        }

        preRendering() {
            this.template_data = {current_sheet_id : this.es_controller.table.template_data.id};
            this.modal_overlay.show();
            super.render();
        }

    };

    module.exports.View = ImportDialogView;

});
